besu/src/main/java/org/hyperledger/besu/controller/MergeBesuControllerBuilder.java

```diff
@@ -73,13 +78,49 @@ protected MiningCoordinator createMiningCoordinator(
                 storageProvider, ScheduleBasedBlockHeaderFunctions.create(protocolSchedule))));
   }
 
+  @Override
+  protected EthProtocolManager createEthProtocolManager(
+      final ProtocolContext protocolContext,
+      final boolean fastSyncEnabled,
+      final TransactionPool transactionPool,
+      final EthProtocolConfiguration ethereumWireProtocolConfiguration,
+      final EthPeers ethPeers,
+      final EthContext ethContext,
+      final EthMessages ethMessages,
+      final EthScheduler scheduler,
+      final List<PeerValidator> peerValidators) {
+
+    EthProtocolManager ethProtocolManager =
+        super.createEthProtocolManager(
+            protocolContext,
+            fastSyncEnabled,
+            transactionPool,
+            ethereumWireProtocolConfiguration,
+            ethPeers,
+            ethContext,
+            ethMessages,
+            scheduler,
+            peerValidators);
+
+    ConsensusContext cc = protocolContext.getConsensusContext(ConsensusContext.class);
+    if (cc instanceof MergeContext) {
+      protocolContext
+          .getConsensusContext(MergeContext.class)
+          .observeNewIsPostMergeState(ethProtocolManager);
+      protocolContext
+          .getConsensusContext(MergeContext.class)
+          .addNewForkchoiceMessageListener(ethProtocolManager);
+    }
+
+    return ethProtocolManager;
+  }
```

Comment:
if we passed in an Optional<PeerFilterAction> or however we encapsulate the merge-specific logic, we wouldn't need to do this kind of introspection on class type.